File: test/arguments/with_range_check_code/cpp98/ChoiceBit4RangeCheckTest.cpp
#include "gtest/gtest.h"

#include "zserio/CppRuntimeException.h"

#include "with_range_check_code/choice_bit4_range_check/ChoiceBit4RangeCheckCompound.h"

namespace with_range_check_code
{
namespace choice_bit4_range_check
{

class ChoiceBit4RangeCheckTest : public ::testing::Test
{
protected:
    void checkChoiceBit4Value(uint8_t value)
    {
        ChoiceBit4RangeCheckCompound choiceBit4RangeCheckCompound;
        const bool selector = true;
        choiceBit4RangeCheckCompound.initialize(selector);
        choiceBit4RangeCheckCompound.setValue(value);
        zserio::BitStreamWriter writer;
        choiceBit4RangeCheckCompound.write(writer);
        size_t writeBufferByteSize;
        const uint8_t* writeBuffer = writer.getWriteBuffer(writeBufferByteSize);
        zserio::BitStreamReader reader(writeBuffer, writeBufferByteSize);
        const ChoiceBit4RangeCheckCompound readChoiceBit4RangeCheckCompound(reader, selector);
        ASSERT_EQ(choiceBit4RangeCheckCompound, readChoiceBit4RangeCheckCompound);
    }

    static const uint8_t    BIT4_LOWER_BOUND = UINT8_C(0);
    static const uint8_t    BIT4_UPPER_BOUND = UINT8_C(15);
};

TEST_F(ChoiceBit4RangeCheckTest, choiceBit4LowerBound)
{
    checkChoiceBit4Value(BIT4_LOWER_BOUND);
}

TEST_F(ChoiceBit4RangeCheckTest, choiceBit4UpperBound)
{
    checkChoiceBit4Value(BIT4_UPPER_BOUND);
}

TEST_F(ChoiceBit4RangeCheckTest, choiceBit4AboveUpperBound)
{
    try
    {
        checkChoiceBit4Value(BIT4_UPPER_BOUND + 1);
        FAIL() << "Actual: no exception, Expected: zserio::CppRuntimeException";
    }
    catch (const zserio::CppRuntimeException& excpt)
    {
        ASSERT_STREQ("Value 16 of ChoiceBit4RangeCheckCompound.value exceeds the range of <0..15>!",
                excpt.what());
    }
}

} // namespace choice_bit4_range_check
} // namespace with_range_check_code

File: test/arguments/with_range_check_code/java/with_range_check_code/ChoiceBit4RangeCheckTest.java
package with_range_check_code;

import static org.junit.jupiter.api.Assertions.*;
import org.junit.jupiter.api.Test;

import with_range_check_code.choice_bit4_range_check.ChoiceBit4RangeCheckCompound;

import java.io.IOException;

import zserio.runtime.ZserioError;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;

public class ChoiceBit4RangeCheckTest
{
    @Test
    public void choiceBit4LowerBound() throws IOException, ZserioError
    {
        checkChoiceBit4Value(BIT4_LOWER_BOUND);
    }

    @Test
    public void choiceBit4UpperBound() throws IOException, ZserioError
    {
        checkChoiceBit4Value(BIT4_UPPER_BOUND);
    }

    @Test
    public void choiceBit4BelowLowerBound() throws IOException, ZserioError
    {
        assertThrows(ZserioError.class, () -> checkChoiceBit4Value((byte)(BIT4_LOWER_BOUND - 1)));
    }

    @Test
    public void choiceBit4AboveUpperBound() throws IOException, ZserioError
    {
        assertThrows(ZserioError.class, () -> checkChoiceBit4Value((byte)(BIT4_UPPER_BOUND + 1)));
    }

    private void checkChoiceBit4Value(byte value) throws IOException, ZserioError
    {
        final boolean selector = true;
        ChoiceBit4RangeCheckCompound choiceBit4RangeCheckCompound = new ChoiceBit4RangeCheckCompound(selector);
        choiceBit4RangeCheckCompound.setValue(value);
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        choiceBit4RangeCheckCompound.write(writer);

        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(writer.toByteArray(),
                writer.getBitPosition());
        final ChoiceBit4RangeCheckCompound readChoiceBit4RangeCheckCompound =
                new ChoiceBit4RangeCheckCompound(reader, selector);
        assertEquals(choiceBit4RangeCheckCompound, readChoiceBit4RangeCheckCompound);
    }

    private static final byte BIT4_LOWER_BOUND = 0;
    private static final byte BIT4_UPPER_BOUND = 15;
}
